STORM-1276: line for line translation of nimbus to java

revans2 opened the pull request:
There are some things I would like to refactor, but this is a function tested translation that I would like to get it, at least as a starting point.

revans2 commented:
All the code is ready if people want to take a look.

dossett commented:
I don't know enough about nimbus or clojure to vote on this PR, but this is an exciting change!

revans2 commented:
@dossett if you want to build the branch and do some manual testing with your own topologies especially looking at the UI that would be helpful.  I did it, but only with word count, and I looked at the UI, but I don't know how exhaustive it was.  So any help you can give would be great.

revans2 commented:
Just fixed the merge conflict, and reran the tests

revans2 commented:
I just rebased on top of the latest master (with a java worker translation)

revans2 commented:
The test failures look unrelated.  It would really be nice if someone could fix the TumblingWindowTest and SlidingWindowTest.  They fail way too often.

ptgoetz commented:
Nice (and a lot of ;) ) work!

Nimbus.java is a little unwieldy, but I understand why. Maybe it can be a target for some refactoring later on.

+1 (Note that I haven't manually tested yet.)

revans2 commented:
@ptgoetz I totally agree on refactoring it.  I did a little as I ported it over, but nothing like it needs.  I also have done some manual testing.

abellina commented:
nimbus.clj, I made it to mkAssignments. Will continue from there later:

- nit: makeBlobCachMap -> makeBlobCacheMap
- change: `numDesiredWorkers < numAssignedWorkers` to `numDesiredWorkers > numAssignedWorkers` in: https://github.com/revans2/incubator-storm/blob/39148edf07f2028c2edcfecc53ea9bfac525988f/storm-core/src/jvm/org/apache/storm/daemon/nimbus/Nimbus.java#L1605
- in getResourcesForTopology: do we care if the values are set? (is_set_cpu for example). Those values are doubles.
- nit: update comment //TODO remove both of swaps below at first opportunity. But yeah "uglyResources" is right :)

srdo commented on storm-core/src/jvm/org/apache/storm/stats/StatsUtil.java:
```diff
@@ -1505,19 +1508,18 @@ public static ComponentPageInfo aggCompExecsStats(
-            List value = (List) entry.getValue();
+        for (Entry<List<? extends Number>, List<Object>> entry : exec2hostPort.entrySet()) {
+            List<? extends Number> key = entry.getKey();
+            List value = entry.getValue();
```
Missing `<Object>`?

srdo commented on storm-core/src/jvm/org/apache/storm/utils/Utils.java:
```diff
@@ -1672,11 +1672,12 @@ public void uncaughtException(Thread thread, Throwable thrown) {
      * @param key The key pointing to the value to be redacted
```
Nit: This method now redacts values of any type, Javadoc should maybe be updated

revans2 replied:
done

revans2 commented on storm-core/src/clj/org/apache/storm/testing.clj:
```diff
@@ -242,11 +297,12 @@
-  (doseq [s @(:supervisors cluster-map)]
-    (.shutdownAllWorkers s nil ReadClusterState/THREAD_DUMP_ON_ERROR)
-    (.close s))
+  (if (:state cluster-map) (.close (:state cluster-map)))
```
In the current code there are multiple "constructors" for the cluster-map.  It was not clear to me that I added state everywhere so I decided to be defensive in my code (especially because I saw some tests that manipulate maps that I though should be internal).  In the next pull request that replaces testing.clj etc all of this gets cleaned up.

revans2 commented on storm-core/src/jvm/org/apache/storm/nimbus/NimbusInfo.java:
```diff
@@ -49,17 +51,17 @@ public static NimbusInfo parse(String nimbusInfo) {
             }
 
-            int port = Integer.parseInt(conf.get(Config.NIMBUS_THRIFT_PORT).toString());
+            int port = Utils.getInt(conf.get(Config.NIMBUS_THRIFT_PORT), 6627);
```
Yes but this helps with unit tests where we don't have to supply configs even when things are mocked out.

revans2 commented on storm-core/src/jvm/org/apache/storm/security/auth/AuthUtils.java:
```diff
@@ -167,15 +175,21 @@ public static IPrincipalToLocal GetPrincipalToLocalPlugin(Map storm_conf) {
     /**
      * Construct a group mapping service provider plugin
-     * @param storm_conf storm configuration
+     * @param conf daemon configuration
```
Yes but it is the daemon configuration vs the topology configuration.

revans2 commented on storm-core/src/clj/org/apache/storm/ui/core.clj:
```diff
@@ -963,11 +963,11 @@
      "port" port
-     "emitted" (Utils/nullToZero (.get_emitted cas))
-     "transferred" (Utils/nullToZero (.get_transferred cas))
+     "emitted" (if-let [em (.get_emitted cas)] em 0)
```
But cas is a thrift object and technically emitted is marked as optional, so being defensive I left the nullToZero in place, but I couldn't make nullToZero work with a Double and return a Long decided to just leave the code the same as before, but go to a pure clojure version that will be replaced with core.clj is translated.

revans2 commented:
I just rebased and I think I have also addressed all of the review comments so far.

erikdw commented on storm-core/src/jvm/org/apache/storm/daemon/nimbus/Nimbus.java:
```diff
@@ -0,0 +1,3640 @@
+    private static Map<String, Map<List<Long>, List<Object>>> computeNewTopoToExecToNodePort(Map<String, SchedulerAssignment> schedAssignments,
+            Map<String, Assignment> existingAssignments) {
+        Map<String, Map<List<Long>, List<Object>>> ret = computeTopoToExecToNodePort(schedAssignments);
+        //Print some useful information
```
Minor nit:  space between `//` and `P`?  Seems *most* of the code puts a space after the comment chars.  But egads, 17% of the existing java code in the storm repo doesn't adhere to that common style practice.  (Consistency being a deposed king it seems.)

```
(storm-reviewing) [STORM-1276] % git grep '//\s' | grep ".java:" | wc -l
    4285
(storm-reviewing) [STORM-1276] % git grep '//\S' | grep ".java:" | grep -v -E "htt(p|ps):" | wc -l
     833
```

erikdw commented on storm-core/src/jvm/org/apache/storm/daemon/nimbus/Nimbus.java:
```diff
@@ -0,0 +1,3729 @@
+        IScheduler scheduler = inimbus == null ? null : inimbus.getForcedScheduler();
+        if (scheduler != null) {
+            LOG.info("Using forced scheduler from INimbus {} {}", scheduler.getClass(), scheduler);
+        } else if (schedClass != null){
```
nit: space between `)` and `{`  (Yes, I'm annoying like this 😉)

erikdw commented on storm-core/src/jvm/org/apache/storm/daemon/nimbus/Nimbus.java:
```diff
@@ -0,0 +1,3729 @@
+    private void updateHeartbeats(String topoId, Set<List<Integer>> allExecutors, Assignment existingAssignment) {
+        LOG.debug("Updating heartbeats for {} {}", topoId, allExecutors);
+        IStormClusterState state = stormClusterState;
+        Map<List<Integer>, Map<String, Object>> executorBeats = StatsUtil.convertExecutorBeats(state.executorBeats(topoId, exis
```
This is not accurate, I'm not sure why @abellina stated this when he changed the code in [c0c72eab](https://github.com/apache/storm/commit/c0c72eab)?  The storm-mesos project *definitely* uses the other arguments:
* [`topologies` being used](https://github.com/mesos/storm/blob/68432a15103cdf00b137ef740730c2e7b7682547/storm/src/main/storm/mesos/schedulers/DefaultScheduler.java#L153)
* [`missingAssignmentTopologies` being used](https://github.com/mesos/storm/blob/master/storm/src/main/storm/mesos/schedulers/DefaultScheduler.java#L152)
* [the call that leads to that code from our class that implements the `INimbus` interface](https://github.com/mesos/storm/blob/68432a15103cdf00b137ef740730c2e7b7682547/storm/src/main/storm/mesos/MesosNimbus.java#L353-L363)

erikdw replied:
nit: inconsistent spacing after `//` between these 2 lines.  Also weird trailing `}`.

I would also say that these comments should be moved up *above* the code where you remove the `deadPorts` from `allPorts`.

erikdw replied:
Comment is referencing the clojure-specific ID of `executor->node+port`.  That is now `topologyToExecutorToNodePort` right?   Although maybe the english-like explanatory style of the clojure ID is preferable?

erikdw replied:
`scratch-topology-id` -> `scratchTopoId`

erikdw replied:
Converting that `mk-assignments` clojure code into Java is pretty impressive; I cannot claim to fully grasp all the logic here, but good job nonetheless!

revans2 replied:
If we want to fix an issue due to STORM-2126 I would propose that we do that on a separate JIRA.  Mostly because STORM-2126 went into the 1.1 line as well as the 2.0 line.  If that goes in first I would then be happy to translate it to java.

revans2 commented:
@erikdw I addressed your nits.  I appreciate the nits because we don't have a coding standard yet.

If we are just down to nits I really would like to get this in.

erikdw commented:
@revans2 : Since I didn't have any substantive blocking commits I'm ok with merging it.  And it's been long enough that I think you're probably ok to merge away.  Any hidden problems will hopefully be surfaced by subsequent testing by others when this is in the mainline.

srdo commented on storm-core/src/jvm/org/apache/storm/daemon/nimbus/Nimbus.java:
```diff
@@ -0,0 +1,3729 @@
+            }
+        } catch (Exception e) {
+            //Yes eat the exception
+            LOG.info("Exception {}", e);
```
Nit: Don't think the {} is necessary

revans2 replied:
The "{}" will make it so the rsult of `e.toString()` is printed instead of the entire stack trace.

srdo commented on storm-core/src/jvm/org/apache/storm/daemon/nimbus/Nimbus.java:
```diff
@@ -0,0 +1,3729 @@
+    private void updateHeartbeats(String topoId, Set<List<Integer>> allExecutors, Assignment existingAssignment) {
+        LOG.debug("Updating heartbeats for {} {}", topoId, allExecutors);
+        IStormClusterState state = stormClusterState;
+        Map<List<Integer>, Map<String, Object>> executorBeats = StatsUtil.convertExecutorBeats(state.executorBeats(topoId, ex
```
Nit: Should maybe be "cancel blob upload exception"

srdo replied:
Not sure this log message fits

srdo replied:
Nit: "ino" -> "info"

srdo commented:
I only really have nitpicks, so I'm good with merging too. Really nice job.

revans2 commented:
I am merging I am still happy to make changes on follow on JIRA.
